src/TypeInference/Pattern.ml

```diff
@@ -140,7 +140,68 @@ let get_ctor_info ~pos ~env (cpath : S.ctor_name S.path S.node) tp =
   | Whnf_Effect _ | Whnf_Effrow _ ->
     failwith "Internal kind error"
 
-let rec check_ctor_named_args ~pos ~env ~scope nps named =
+(** Introduce all named types and named arguments into the environment.
+  The type variables are refreshed, so the function returns a renaming. *)
+let open_named ~pos ~public env targs named =
+  let ((env, sub), targs) =
+    List.fold_left_map
+      (fun (env, sub) (name, x) ->
+        let (env, y) =
+          let kind = T.TVar.kind x in
+          match name with
+          | T.TNAnon   -> Env.add_anon_tvar env kind
+          | T.TNEffect ->
+            assert (T.Kind.is_effect kind);
+            Env.add_the_effect env
+          | T.TNVar x  -> Env.add_tvar ~public env x kind
+        in
+        ((env, T.Subst.rename_to_fresh sub x y), y))
+      (env, T.Subst.empty)
+      targs
+  in
+  let named = List.map (T.NamedScheme.subst sub) named in
+  let (env, named) =
+    List.fold_left_map
+      (fun env (name, sch) ->
+        let (env, x) =
+          match name with
+          | T.NLabel -> Env.add_the_label_sch env sch
+          | T.NVar x -> Env.add_poly_var ~public env x sch
+          | T.NImplicit n -> Env.add_poly_implicit ~public env n sch ignore
+          | T.NMethod   n ->
+            let owner = TypeUtils.method_owner_of_scheme ~pos ~env sch in
+            Env.add_poly_method ~public env owner n sch
+        in
+        (env, T.{ pos; data = PVar(x, sch) }))
+      env
+      named in
+  (env, targs, named, sub)
+
+let rec check_ctor_named ~pos ~env ~scope
+    (cpath : S.ctor_name S.path S.node) (ctor : T.ctor_decl) named =
+  match named with
+  | S.CNParams(targs, nps) ->
+    Uniqueness.check_named_type_arg_uniqueness targs;
+    let targs = List.map tr_named_type targs in
+    let (env, scope, sub2, tvars) =
+      check_ctor_type_args ~pos:cpath.pos ~env ~scope ~sub:T.Subst.empty
+        targs ctor.ctor_targs in
+    let ctor_named = List.map (T.NamedScheme.subst sub2) ctor.ctor_named in
+    Uniqueness.check_named_pattern_uniqueness nps;
+    let nps = List.map (tr_named_pattern env) nps in
+    let (env, bn1, ps1) =
+      check_ctor_named_args ~pos ~env ~scope nps ctor_named in
+    (env, scope, sub2, tvars, ps1, bn1)
+  | S.CNModule modname ->
+    (* TODO: This may seem inconsistent with the other case as implicits aren't
+       introduced to the current namespace, just the provided module name. *)
```

Comment:
Indeed. This should be discussed and we should find a general policy of implicit introduction of variables. Before that, we can live with what is done.